File: android/src/main/java/com/auth0/react/Passwordless.kt
package com.auth0.react

import com.auth0.android.Auth0
import com.auth0.android.authentication.AuthenticationAPIClient
import com.auth0.android.authentication.AuthenticationException
import com.auth0.android.authentication.passwordless.DeliveryMethod
import com.auth0.android.result.Credentials
import com.auth0.android.result.PasswordlessChallenge
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.WritableNativeMap

class Passwordless(
    auth0: Auth0,
    useDPoP: Boolean,
    reactContext: ReactApplicationContext
) {

    private val client: AuthenticationAPIClient = AuthenticationAPIClient(auth0).also {
        if (useDPoP) {
            it.useDPoP(reactContext)
        }
    }

    fun challengeWithEmail(
        email: String,
        connection: String,
        allowSignup: Boolean = false,
        promise: Promise
    ) {
        client.passwordlessClient()
            .challengeWithEmail(email, connection, allowSignup)
            .start(object : com.auth0.android.callback.Callback<PasswordlessChallenge, AuthenticationException> {
                override fun onSuccess(result: PasswordlessChallenge) {
                    val map = WritableNativeMap().apply {
                        putString("authSession", result.authSession)
                    }
                    promise.resolve(map)
                }

                override fun onFailure(error: AuthenticationException) {
                    promise.reject("PASSWORDLESS_CHALLENGE_FAILED", error.getDescription(), error)
                }
            })
    }

    fun challengeWithPhoneNumber(
        phoneNumber: String,
        connection: String,
        deliveryMethod: String,
        allowSignup: Boolean = false,
        promise: Promise
    ) {
        val method = if (deliveryMethod == "voice") DeliveryMethod.VOICE else DeliveryMethod.TEXT

        client.passwordlessClient()
            .challengeWithPhoneNumber(phoneNumber, connection, method, allowSignup)
            .start(object : com.auth0.android.callback.Callback<PasswordlessChallenge, AuthenticationException> {
                override fun onSuccess(result: PasswordlessChallenge) {
                    val map = WritableNativeMap().apply {
                        putString("authSession", result.authSession)
                    }
                    promise.resolve(map)
                }

                override fun onFailure(error: AuthenticationException) {
                    promise.reject("PASSWORDLESS_CHALLENGE_FAILED", error.getDescription(), error)
                }
            })
    }

    fun loginWithOTP(
        authSession: String,
        otp: String,
        audience: String?,
        scope: String?,
        promise: Promise
    ) {
        val finalScope = if (scope.isNullOrBlank()) "openid profile email" else scope
        val finalAudience = audience?.trim()?.ifEmpty { null }

        val request = client.passwordlessClient()
            .loginWithOTP(PasswordlessChallenge(authSession), otp)
        finalAudience?.let { request.setAudience(it) }
        request.setScope(finalScope)

        request.start(object : com.auth0.android.callback.Callback<Credentials, AuthenticationException> {
            override fun onSuccess(result: Credentials) {
                promise.resolve(CredentialsParser.toMap(result))
            }

            override fun onFailure(error: AuthenticationException) {
                promise.reject("PASSWORDLESS_LOGIN_FAILED", error.getDescription(), error)
            }
        })
    }
}

File: ios/Passwordless.swift
import Auth0
import Foundation

@objc
public class A0Passwordless: NSObject {

    private let client: Authentication

    @objc public init(clientId: String, domain: String, useDPoP: Bool) {
        var client = Auth0.authentication(clientId: clientId, domain: domain)
        if useDPoP {
            client = client.useDPoP()
        }
        self.client = client
    }

    @objc(challengeWithEmail:connection:allowSignup:resolve:reject:)
    public func challengeWithEmail(email: String, connection: String, allowSignup: Bool, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        client.passwordlessChallenge(
            email: email,
            connection: connection,
            allowSignup: allowSignup
        ).start { result in
            switch result {
            case .success(let challenge):
                resolve(["authSession": challenge.authSession])
            case .failure(let error):
                reject("PASSWORDLESS_CHALLENGE_FAILED", error.localizedDescription, error)
            }
        }
    }

    @objc(challengeWithPhoneNumber:connection:deliveryMethod:allowSignup:resolve:reject:)
    public func challengeWithPhoneNumber(phoneNumber: String, connection: String, deliveryMethod: String, allowSignup: Bool, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let method = DeliveryMethod(rawValue: deliveryMethod) ?? .text

        client.passwordlessChallenge(
            phoneNumber: phoneNumber,
            connection: connection,
            deliveryMethod: method,
            allowSignup: allowSignup
        ).start { result in
            switch result {
            case .success(let challenge):
                resolve(["authSession": challenge.authSession])
            case .failure(let error):
                reject("PASSWORDLESS_CHALLENGE_FAILED", error.localizedDescription, error)
            }
        }
    }

    @objc(loginWithOTP:otp:audience:scope:resolve:reject:)
    public func loginWithOTP(authSession: String, otp: String, audience: String?, scope: String?, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let audienceValue = audience?.isEmpty == true ? nil : audience
        let finalScope = scope?.isEmpty == true ? "openid profile email" : (scope ?? "openid profile email")

        // PasswordlessChallenge has no public initializer, but it is Codable with
        // `authSession` mapped to the `auth_session` JSON key. Reconstruct it by decoding.
        guard let challengeData = try? JSONSerialization.data(withJSONObject: ["auth_session": authSession]),
              let challenge = try? JSONDecoder().decode(PasswordlessChallenge.self, from: challengeData) else {
            reject("PASSWORDLESS_LOGIN_FAILED", "Invalid passwordless challenge", nil)
            return
        }

        client.login(
            otp: otp,
            challenge: challenge,
            audience: audienceValue,
            scope: finalScope
        ).start { result in
            switch result {
            case .success(let credentials):
                resolve(credentials.asDictionary())
            case .failure(let error):
                reject("PASSWORDLESS_LOGIN_FAILED", error.localizedDescription, error)
            }
        }
    }
}
